### PR TITLE
Feature/simplify passed info

### PR DESCRIPTION
**Pull request by jrfnl**
Tested reasonably extensively

### Improve plugin registration method

- register plugins with the `$slug` as associative array key
- ensure that all plugins have defaults registered for the keys not passed
- add `$has_forced_activation` and `$has_forced_deactivation` properties to the TGMPA class which are set from the plugin registration method and used to determine whether to add the relevant actions
- add `$sort_order` property which holds the plugin names for use in sorting the `$plugins` property for internal use
- as - no matter what - we'll always need the `file_path`, we might as well also set it in the registration method and get rid of the repeated calls to `$this->populate_file_path()`
- also made the `populate_file_path()` a little bit more flexible so it can be used in a leaner manner for those cases where it's still needed

### Simplify the table generation, install and activate code

- retrieve most information we need from our `$plugins` array rather than pass it through URL or `$_POST`
- properly sanitize and validate received information from URL or `$_POST`
- leverage the use of the WP_filesystem API
- show an appropriate error message if user tries to activate an already active plugin
- deprecate the `TGMPA_List_Table::_get_plugin_data_from_name()` method which is no longer needed as we have access to the `TGM_Plugin_Activation::_get_plugin_data_from_name()` method
- Fix unauthorized issue on bulk installs when not using direct file access caused by #227
- Only prepare some variables and update recently activated plugins option when we actually need to

Also:
- abstract the regexes which are used a number of times to class constants
- combine multiple `isset()`'s if used within the same if statement
- don't nest what doesn't need nesting
- unset loop variables if the function doesn't return straight after

### CODE REVIEW

File: class-tgm-plugin-activation.php [hunk 1]
```diff
@@ -758,17 +775,17 @@ public function notices() {
 					$install_link = true; // We need to display the 'install' action link.
 					$install_link_count++; // Increment the install link count.
 					if ( current_user_can( 'install_plugins' ) ) {
-						if ( isset( $plugin['required'] ) && $plugin['required'] ) {
-							$message['notice_can_install_required'][] = $plugin['name'];
+						if ( true === $plugin['required'] ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: How about if the user has registered slightly incorrect as a string:

`required => 'true',`

? We should allow a little leeway for 2.5, and then make it stricter for 3.0.

File: class-tgm-plugin-activation.php [hunk 2]
```diff
@@ -726,28 +752,19 @@ public function notices() {
 			$activate_link       = false;   // Set to false, change to true in loop if conditions exist, used for action link 'activate'.
 			$activate_link_count = 0;       // Used to determine plurality of activate action link text.
 
-			foreach ( $this->plugins as $plugin ) {
+			foreach ( $this->plugins as $slug => $plugin ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Can this section of deciding which messages need showing be refactored out?

File: class-tgm-plugin-activation.php [hunk 3]
```diff
@@ -802,25 +820,24 @@ public function notices() {
 				}
 
 				// Grab all plugin names.
-				foreach ( $message as $type => $plugin_groups ) {
-					$linked_plugin_groups = array();
-
-					// Count number of plugins in each message group to calculate singular/plural message.
-					$count = count( $plugin_groups );
+				foreach ( $message as $type => $plugin_group ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Can this section of looping through messages and showing them be refactored out?

File: class-tgm-plugin-activation.php [hunk 4]
```diff
@@ -828,22 +845,25 @@ public function notices() {
...
+							$linked_plugins[] = esc_html( $this->plugins[ $plugin_slug ]['name'] ); // No hyperlink.
 						}
 					}
+					unset( $plugin_slug, $url );
 
-					$last_plugin = array_pop( $plugin_groups ); // Pop off last name to prep for readability.
-					$imploded    = empty( $plugin_groups ) ? '<em>' . $last_plugin . '</em>' : '<em>' . ( implode( ', ', $plugin_groups ) . '</em> and <em>' . $last_plugin . '</em>' );
+					$count       = count( $plugin_group );
+					$last_plugin = array_pop( $linked_plugins ); // Pop off last name to prep for readability.
+					$imploded    = empty( $linked_plugins ) ? '<em>' . $last_plugin . '</em>' : '<em>' . ( implode( ', ', $linked_plugins ) . '</em> ' . esc_html__( 'and', 'tgmpa' ) . ' <em>' . $last_plugin . '</em>' );
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: How about a small private/protected function that returns a string in `<em>`??

The comma between each plugin will be italicised, so any semantics the emphasis element is meant to have per plugin, is not applicable.

The `esc_html__()` should be `esc_html_x()` so that some context can be given to translators about what that "and" is for.

File: class-tgm-plugin-activation.php (hunk 2 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Should be done in a separate PR, is a different issue. (and I'm preparing a number of those already, mainly focussed on abstracting (nearly) duplicate code to functions)

File: class-tgm-plugin-activation.php (hunk 3 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Should be done in a separate PR, is a different issue. (and I'm preparing a number of those already, mainly focussed on abstracting (nearly) duplicate code to functions)

File: class-tgm-plugin-activation.php [hunk 5]
```diff
@@ -1029,6 +1080,10 @@ public function populate_file_path() {
 		 */
 		protected function _get_plugin_basename_from_slug( $slug ) {
 
+			if ( ! function_exists( 'get_plugins' ) ) {
+				require_once ABSPATH . 'wp-admin/includes/plugin.php';
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Is there a `WP_ADMIN` constant?

File: class-tgm-plugin-activation.php (hunk 1 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: I've added some type casting to the registration method now to account for this. Using some of the filter/filter emulate code I pointed to elsewhere from the WPSEO plugin. Not giving credit to them as I wrote it ;-)

File: class-tgm-plugin-activation.php [hunk 6]
```diff
@@ -1286,25 +1337,25 @@ protected function _gather_plugin_data() {
...
-
-				$external_url = $this->_get_plugin_data_from_name( $plugin['name'], 'external_url' );
-				$source       = $this->_get_plugin_data_from_name( $plugin['name'], 'source' );
+				$table_data[ $i ]['slug']             = $slug;
 
-				if ( $external_url && preg_match( '|^http(s)?://|', $external_url ) ) {
-					$table_data[ $i ]['plugin'] = '<strong><a href="' . esc_url( $external_url ) . '" target="_blank">' . esc_html( $plugin['name'] ) . '</a></strong>';
-				}
-				elseif ( ! $source || preg_match( '|^http://wordpress.org/extend/plugins/|', $source ) ) {
+				if ( ! empty( $plugin['external_url'] ) && preg_match( TGM_Plugin_Activation::EXT_REPO_REGEX, $plugin['external_url'] ) ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Since this condition, and the elseif condition below are already being used around line 829, can you create public boolean methods in the main class, and have both conditions here reference each method?
i.e. (as a quick example)
~~~php
// TGM_Plugin_Activation class
public function is_valid_external_source( $external_url ) {
    return ! empty( $external_url ) && preg_match( self::EXT_REPO_REGEX, $external_url );
}
~~~

~~~php
// Used here as:
if ( $this->tgmpa->is_valid_external_source( $plugin['external_url'] ) {
~~~

The elseif below might be `is_valid_repo_source()`, and the if / elseif in the main plugin (line ~829) would be called as `if ( $this->is_valid_external_source( $this->plugins[ $plugin_slug ]['external_url'] )` etc.

File: class-tgm-plugin-activation.php [hunk 7]
```diff
@@ -1286,25 +1337,25 @@ protected function _gather_plugin_data() {
...
+					$table_data[ $i ]['plugin'] = sprintf(
+						'<strong><a href="%1$s" target="_blank">%2$s</a></strong>',
+						esc_url( $plugin['external_url'] ),
+						esc_html( $plugin['name'] )
+					);
+				} elseif ( 'repo' === $plugin['source'] || preg_match( TGM_Plugin_Activation::WP_REPO_REGEX, $plugin['source'] ) ) {
 					$url = add_query_arg(
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: This `$url` generation could be refactored  with the earlier one too.

File: class-tgm-plugin-activation.php (hunk 4 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: I like idea, but again, should be done in a separate PR, not the same issue.

File: class-tgm-plugin-activation.php [hunk 8]
```diff
@@ -1862,6 +1816,26 @@ public function prepare_items() {
...
 
+
+		/* *********** DEPRECATED METHODS *********** */
+
+		/**
+		 * Retrieve plugin data, given the plugin name.
+		 *
+		 * @since      2.2.0
+		 * @deprecated 2.5.0
+		 * @deprecated use TGM_Plugin_Activation::_get_plugin_data_from_name()
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: This line is redundant.

File: class-tgm-plugin-activation.php [hunk 9]
```diff
@@ -1501,18 +1528,12 @@ public function column_plugin( $item ) {
...
-			if ( __( 'Private Repository', 'tgmpa' ) === $item['source'] ) {
-				// Escape external URLs
-				$plugin_url = esc_url( $plugin_url );
-			}
-			elseif ( __( 'Pre-Packaged', 'tgmpa' ) === $item['source'] ) {
-				// Encode file path for use in attribute
-				$plugin_url = urlencode( $this->tgmpa->default_path . $plugin_url );
-			}
-
-			$value = $item['file_path'] . ',' . $plugin_url . ',' . $item['sanitized_plugin'];
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Was this just a hacky way to get all of the information into `$_POST`?

File: class-tgm-plugin-activation.php (hunk 5 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: There is, but it's a boolean, so not a path to admin ;-)

File: class-tgm-plugin-activation.php [hunk 10]
```diff
@@ -1587,176 +1607,113 @@ public function process_bulk_actions() {
...
-				}
-				// Looks like the user can use the direct method, take from $_POST.
-				elseif ( isset( $_POST['plugin'] ) ) {
-					$plugins = (array) $_POST['plugin'];
-				}
-				// Nothing has been submitted.
-				else {
-					$plugins = array();
-				}
+				if ( isset( $_POST['plugin'] ) && ! empty( $_POST['plugin'] ) ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: I thought you said `empty()` already checks for `isset()`?

File: class-tgm-plugin-activation.php (hunk 5 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: https://wordpress.org/plugins/debug-bar-constants/ FTW :sunglasses:

File: class-tgm-plugin-activation.php [hunk 11]
```diff
@@ -1587,176 +1607,113 @@ public function process_bulk_actions() {
...
-							$plugins_to_install[] = $value;
-							$plugin_installs[]    = $value;
+					
+					// Sanitize the received input
+					$plugins_to_install = array_map( 'sanitize_title', $plugins_to_install );
+
+					// Validate the received input
+					foreach ( $plugins_to_install as $key => $slug ) {
+						// Check if the plugin was registered with TGMPA and remove if not
+						if ( ! isset( $this->tgmpa->plugins[ $slug ] ) ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Is this logic correct?

"If NOT a registered plugin, ignore this one we want to install"?

File: class-tgm-plugin-activation.php (hunk 6 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: I'm abstracting this (whole block of code) in a different way in one of the new branches

File: class-tgm-plugin-activation.php (hunk 7 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Dito

File: class-tgm-plugin-activation.php (hunk 8 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Removed

File: class-tgm-plugin-activation.php (hunk 9 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Hell yeah

File: class-tgm-plugin-activation.php (hunk 10 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Good catch

File: class-tgm-plugin-activation.php [hunk 12]
```diff
@@ -1587,176 +1607,113 @@ public function process_bulk_actions() {
...
-				// Loop through each plugin to install and try to grab information from WordPress API, if not create 'tgmpa-empty' scalar.
-				$i = 0;
-				foreach ( $plugin_installs as $plugin ) {
-					$temp = plugins_api( 'plugin_information', array( 'slug' => $plugin, 'fields' => array( 'sections' => false ) ) );
+				// Prepare the data for validated plugins for the install
+				foreach ( $plugins_to_install as $slug ) {
+					$name   = $this->tgmpa->plugins[ $slug ]['name'];
+					$source = $this->tgmpa->plugins[ $slug ]['source'];
+
+					if ( 'repo' === $source || preg_match( TGM_Plugin_Activation::WP_REPO_REGEX, $source ) ) {
```
Comment (GaryJones) [inline class-tgm-plugin-activation.php]: Here's another instance where the earlier mentioned refactoring could be used.

File: class-tgm-plugin-activation.php (hunk 11 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: Logic is correct. We only want to deal with plugins which were registered in TGMPA, if a plugin was not registered by a theme or plugin using TGMPA and still somehow managed to get into the $_POST array, we ignore it as it's probably some sort of hack.

File: class-tgm-plugin-activation.php (hunk 12 above)
Comment (jrfnl) [inline class-tgm-plugin-activation.php]: See above

**Comment by GaryJones** [comment]
Phew! Excellent stuff. I realise some of my suggestions may well be other PRs once this is merged, but wanted to mention them to get them on your radar if they weren't already.
